lib/mix/lib/mix/tasks/test.ex

```diff
@@ -679,14 +720,71 @@ defmodule Mix.Tasks.Test do
     end
   end
 
-  defp display_warn_test_pattern(test_files, test_pattern, matched_test_files, warn_test_pattern) do
-    files = Mix.Utils.extract_files(test_files, warn_test_pattern) -- matched_test_files
+  defp directly_included_test_files(files) do
+    Enum.filter(files, fn path -> :elixir_utils.read_file_type(path) == {:ok, :regular} end)
+  end
 
-    for file <- files do
-      Mix.shell().info(
-        "warning: #{file} does not match #{inspect(test_pattern)} and won't be loaded"
-      )
-    end
+  defp classify_test_files(potential_test_files, project) do
+    test_load_filters = project[:test_load_filters] || [~r/.*_test\.exs$/]
+    elixirc_paths = project[:elixirc_paths] || []
+
+    # ignore any _helper.exs files and files that are compiled (test support files)
+    test_warn_filters =
+      Keyword.get_lazy(project, :test_warn_filters, fn ->
+        [
+          ~r/.*_helper\.exs$/,
+          fn file -> Enum.any?(elixirc_paths, &String.starts_with?(file, &1)) end
```

Comment:
This is not going to work if you give a full path to test: `mix test $PWD/test`

Reply:
It may also not work on Windows because of `/` and `\\`.

Reply:
@josevalim it does work for for full paths, since they are resolved to a relative path first since https://github.com/elixir-lang/elixir/pull/13258, but I'll need to create a Windows VM to test Windows 😃

Reply:
Fantastic. For windows, you need to confirm a file given as test\foo will still work. I am not sure if Path.wildcard will normalize all names. And we need to document all patterns use / slashes (if that’s actually the case).

Reply:
Looks like all paths are properly normalized. Everything I tried works as expected:

```
PS C:\Users\Steffen\test> mix test
WINDOWS! file: test/foo.ex; elixirc path is lib
WINDOWS! file: test/foo.ex; elixirc path is test/support
WINDOWS! file: test/sub/foo.ex; elixirc path is lib
WINDOWS! file: test/sub/foo.ex; elixirc path is test/support
WINDOWS! file: test/support/foo.ex; elixirc path is lib
WINDOWS! file: test/support/foo.ex; elixirc path is test/support
warning: the following files do not match any of the configured `:test_load_filters` / `:test_warn_filters`:

test/foo.ex
test/sub/foo.ex

This might indicate a typo in a test file name (for example, using "foo_tests.exs" instead of "foo_test.exs").

You can adjust which files trigger this warning by configuring the `:test_warn_filters` option in your
Mix project's configuration. To disable the warning entirely, set that option to false.

For more information, run `mix help test`.

Running ExUnit with seed: 691107, max_cases: 24

...
Finished in 0.05 seconds (0.00s async, 0.05s sync)
1 doctest, 2 tests, 0 failures
```

```
PS C:\Users\Steffen\test> mix test test\sub\
WINDOWS! file: test/sub/foo.ex; elixirc path is lib
WINDOWS! file: test/sub/foo.ex; elixirc path is test/support
warning: the following files do not match any of the configured `:test_load_filters` / `:test_warn_filters`:

test/sub/foo.ex

This might indicate a typo in a test file name (for example, using "foo_tests.exs" instead of "foo_test.exs").

You can adjust which files trigger this warning by configuring the `:test_warn_filters` option in your
Mix project's configuration. To disable the warning entirely, set that option to false.

For more information, run `mix help test`.

Running ExUnit with seed: 766859, max_cases: 24

.
Finished in 0.04 seconds (0.00s async, 0.04s sync)
1 test, 0 failures
```

I added the `WINDOWS!` line for debugging.

Reply:
Fantastic. Just double checking: you used Powershell instead of WSL? That's important because maybe it would perform Unix normalization on WSL.

If that's good, let's document that all paths given to filters use forward slashes and I believe we should be good to go once the other comments are addressed. Sorry that this issue seems to be quite more work than initially appeared. :)